leopard.go

```diff
@@ -1,54 +1,63 @@
-//go:build leopard
-
-// Note that if the build tag leopard is used, liblibleopard.a
-// has to be present where the linker will find it.
-// Otherwise go-leopard won't build.
 package rsmt2d
 
-import "github.com/celestiaorg/go-leopard"
+import (
+	"github.com/klauspost/reedsolomon"
+)
 
-var _ Codec = leoRSFF8Codec{}
-var _ Codec = leoRSFF16Codec{}
+var _ Codec = leoRSCodec{}
 
 func init() {
-	registerCodec(LeopardFF8, newLeoRSFF8Codec())
-	registerCodec(LeopardFF16, newLeoRSFF16Codec())
-}
-
-type leoRSFF8Codec struct{}
-
-func (l leoRSFF8Codec) Encode(data [][]byte) ([][]byte, error) {
-	return leopard.Encode(data)
+	registerCodec(Leopard, newLeoRSCodec())
 }
 
-func (l leoRSFF8Codec) Decode(data [][]byte) ([][]byte, error) {
-	half := len(data) / 2
-	return leopard.Decode(data[:half], data[half:])
+type leoRSCodec struct {
+	encoderCache *doubleCache[reedsolomon.Encoder]
```

Comment:
@rahulghangas, have you used the [`sync.Pool`](https://pkg.go.dev/sync#Pool) primitive before? It seems like the purpose of the pool exactly matches the cache's goal. If not, there are various articles on the internet explaining the primitive. You can also check how @klauspost's reedsolomon library uses it for efficient reusage of allocated byte buffers.

Using Pool will give use:
* Less code to maintain and reinvent
* Non-blocking access to the pooled caches
	* The cache has a mutex which will harm performance for the parallelization of multiple EDS computes. The more, the worst it gets. 
* Does not use `reflect` pkg, which is slow in the way the cache uses it(`TypeOf`)

Reply:
Yes, ideally we should be using the same encoder per `dataLen`, and that is indeed the optimization here (insteda of creating a new encoder each time)